### PR TITLE
fixes #928 - Round-trip for apoc.export.csv.all

### LINKED ISSUE #928: Round-trip for apoc.export.csv.all
It would be good if the file generated by apoc.export.csv.all could be imported into an empty database using the neo4j-admin import command without modification to the generated file.

### PR DESCRIPTION
**Pull request by AngeloBusato**
Fixes #928

Round-trip for apoc.export.csv.all

## Proposed Changes (Mandatory)

A brief list of proposed changes in order to fix the issue:
  - Add the the config to export the csv for the adminImportTool

### CODE REVIEW

File: src/main/java/apoc/export/csv/CsvFormat.java [hunk 1]
```diff
@@ -120,6 +129,111 @@ public void writeAll(SubGraph graph, Reporter reporter, ExportConfig config, CSV
...
+        headerNode.addAll(keyTypes.entrySet().stream().map(stringClassEntry -> formatHeader(stringClassEntry)).collect(Collectors.toCollection(LinkedHashSet::new)));
+        headerNode.add(":LABEL");
+        return headerNode;
+    }
+
+    private Set<String> generateHeaderRelationship(Map.Entry<RelationshipType, List<Relationship>> entrySet) {
+        Set<String> headerNode = new LinkedHashSet<>();
+        Map<String,Class> keyTypes = new LinkedHashMap<>();
+        entrySet.getValue().forEach(relationship -> updateKeyTypes(keyTypes, relationship));
+        headerNode.addAll(keyTypes.entrySet().stream().map(stringClassEntry -> formatHeader(stringClassEntry)).collect(Collectors.toCollection(LinkedHashSet::new)));
```
Comment (jexp) [inline src/main/java/apoc/export/csv/CsvFormat.java]: stringClassEntry ??

File: src/main/java/apoc/export/csv/CsvFormat.java [hunk 2]
```diff
@@ -120,6 +129,111 @@ public void writeAll(SubGraph graph, Reporter reporter, ExportConfig config, CSV
...
+        headerNode.add(":START_ID");
+        headerNode.add(":END_ID");
+        headerNode.add(":TYPE");
+        return headerNode;
+    }
+
+    private void writeRow(ExportConfig config, FileManagerFactory.ExportCypherFileManager writer, Set<String> headerNode, List<List<String>> rows, String name) {
```
Comment (jexp) [inline src/main/java/apoc/export/csv/CsvFormat.java]: nodeHeader

File: src/main/java/apoc/export/csv/CsvFormat.java [hunk 3]
```diff
@@ -120,6 +129,111 @@ public void writeAll(SubGraph graph, Reporter reporter, ExportConfig config, CSV
...
+        try (PrintWriter pw = writer.getPrintWriter(name);
+             CSVWriter csvWriter = getCsvWriter(pw, config)) {
+            if (config.isSeparateHeader()) {
+                try (PrintWriter pwHeader = writer.getPrintWriter("header." + name)) {
+                    CSVWriter csvWriterHeader = getCsvWriter(pwHeader, config);
+                    csvWriterHeader.writeNext(headerNode.toArray(new String[0]), false);
```
Comment (jexp) [inline src/main/java/apoc/export/csv/CsvFormat.java]: can we use a reusable string array of the correct size for all these writeNext? in this method

File: src/main/java/apoc/export/csv/CsvFormat.java (hunk 2 above)
Comment (jexp) [inline src/main/java/apoc/export/csv/CsvFormat.java]: Don#t use a Set for nodeHeader, but a list or array

File: src/main/java/apoc/export/csv/ExportCSV.java [hunk 4]
```diff
@@ -80,40 +76,50 @@ public ExportCSV() {
...
+        checkConfig(config);
         Map<String,Object> params = config == null ? Collections.emptyMap() : (Map<String,Object>)config.getOrDefault("params", Collections.emptyMap());
         Result result = db.execute(query,params);
 
         String source = String.format("statement: cols(%d)", result.columns().size());
         return exportCsv(fileName, source,result,config);
     }
 
+    private void checkConfig(@Name("config") Map<String, Object> config) {
+        if (config != null && config.containsKey("neo4jImport")) throw new RuntimeException("Only apoc.export.all can have the `neo4jAdmin` config");
```
Comment (jexp) [inline src/main/java/apoc/export/csv/ExportCSV.java]: doesn't align with docs `neo4jAdminTool` and should be changed to `bulkImport` anyway.

File: src/main/java/apoc/export/cypher/FileManagerFactory.java [hunk 5]
```diff
@@ -14,79 +14,111 @@
...
 
+        StringWriter getStringWriter(String type);
+
         String drain(String type);
+
+        String getFileName();
     }
 
-    private static class PhysicalExportCypherFileManager implements ExportCypherFileManager {
+    private static class StringExportCypherFileManager implements ExportCypherFileManager {
```
Comment (jexp) [inline src/main/java/apoc/export/cypher/FileManagerFactory.java]: why was the order of theses classes changed?

Comment (jexp) [reply src/main/java/apoc/export/cypher/FileManagerFactory.java]: please put them back so it's possible to see the actual changes.

File: src/main/java/apoc/export/cypher/FileManagerFactory.java [hunk 6]
```diff
@@ -14,79 +14,111 @@
...
+            return new StringExportCypherFileManager(separatedFiles);
+        }
+
+        int indexOfDot = fileName.lastIndexOf(".");
+        String fileType = fileName.substring(indexOfDot + 1);
+        return new PhysicalExportFileManager(fileType, fileName, separatedFiles);
     }
 
-    interface ExportCypherFileManager {
+    public interface ExportCypherFileManager {
```
Comment (jexp) [inline src/main/java/apoc/export/cypher/FileManagerFactory.java]: Remove Cypher from the name, make it a top level interface.

File: src/main/java/apoc/export/util/ExportConfig.java [hunk 7]
```diff
@@ -26,6 +25,7 @@
 
     private int batchSize = DEFAULT_BATCH_SIZE;
     private boolean silent = false;
+    private boolean neo4jImport = false;
```
Comment (jexp) [inline src/main/java/apoc/export/util/ExportConfig.java]: bulkImport

File: src/main/java/apoc/export/csv/ExportCSV.java (hunk 4 above)
Comment (jexp) [reply src/main/java/apoc/export/csv/ExportCSV.java]: Should use exportConfig

File: src/main/java/apoc/export/util/Neo4jAdminImportUtil.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
+package apoc.export.util;
+
+import org.neo4j.values.storable.DurationValue;
+import org.neo4j.values.storable.PointValue;
+
+import java.time.*;
+import java.util.HashMap;
+import java.util.Map;
+
+public class Neo4jAdminImportUtil {
```
Comment (jexp) [inline src/main/java/apoc/export/util/Neo4jAdminImportUtil.java]: BulkImportUtil

File: src/main/java/apoc/export/util/Neo4jAdminImportUtil.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+    private static Map<Class<?>, String> allowedMapping = new HashMap(){{
```
Comment (jexp) [inline src/main/java/apoc/export/util/Neo4jAdminImportUtil.java]: did you check the odd formatting for datetime and spatial that are required? see the docs.

Comment (conker84) [reply src/main/java/apoc/export/util/Neo4jAdminImportUtil.java]: the whole list is taken from here:
https://neo4j.com/docs/operations-manual/current/tools/import/file-header-format/#import-tool-header-format-properties

**Comment by jexp** [comment]
Thanks a lot!

**Pull request merged**